csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -137,43 +136,44 @@ public ValueTask SetAdapterDirectProxyAsync(string adapterId, IObjectPrx? proxy,
                     return null;
                 }
 
-                IObjectPrx prx = _wellKnownProxy.Clone(IObjectPrx.Factory, identity: identity);
-
-                var adapterIds = new List<string>();
-                foreach (KeyValuePair<string, HashSet<string>> entry in _replicaGroups)
+                foreach ((string key, HashSet<string> ids) in _replicaGroups)
                 {
                     try
                     {
-                        prx.Clone(adapterId: entry.Key).IcePing();
-                        adapterIds.Add(entry.Key);
+
+                        IObjectPrx prx = _adapters[ids.First()];
```

Comment:
This does not look correct. The previous version was iterating over _all_ adapters (first all referenced by replica groups, and if no replica group, all regular adapters), gathered all reachable adapters then returned one at random. There is no more randomness here.

Reply:
The previous behavior looked for the Id in all replica-groups and all stand alone adapters and the return one at random.

If we want to keep this behavior we should collect proxies, as different adapters can be using different protocols, and then return a random proxy.  But in Ice object model an Identity should be unique so doesn't make sense that the same ID is part of several replica groups, or adapters.

Reply:
I assume the LocatorRegistry can have several distinct replica groups registered with it, and the desired well-known object is in one of these replica groups. We should indeed break out of the for loop once we've found the object in one replica group.

Then, it is possible (though unlikely) for some of the members of the "good" replica group to be down, hence we check (ping) each adapter ID and select the result at random from one of the adapters that is reachable.

Overall I think the old version was not quite correct, but the new one is not correct either!

Reply:
The adapterId here is `adapterId: key` is the same for all members I think what is not correct is that Clone doesn't clear the endpoints, previously the returned proxy has no endpoints.

Reply:
Actually setting the `adapterId` clears the endpoints so the Clone is just fine

Reply:
Why are we retrieving prx here? Just for its protocol and encoding?

For clearLocator/clearRouter, should we just clear these when we save the proxy?

Reply:
> Why are we retrieving prx here? Just for its protocol and encoding?

Yes, we previously Clone a proxy that always use Ice1 protocol

> For clearLocator/clearRouter, should we just clear these when we save the proxy?

Yes I will fix that